### PR TITLE
ensure the side which enbale SO_LINGER and call showdownOutput to start TCP half-closure in fin_wait2 state can still receive and process the data which is send by another side in the close_wait state

### LINKED ISSUE #11981: [bug] when SO_LINGGER is used in the TCP half-closed scenario, the client in the fin_wait2 state cannot receive the data sent from the server which is in the close_wait state
SO_LINGERHalfClosureBug

as shown in the picture above。

**when `SO_LINGGER` is used in the client side**, and client call `NioSocketChannel#shutdownOutput()` to **half close** the tcp connection。

then server side `OP_READ` is active，server begin to `shutdowninput` and `triger ChannelInputShutdownEvent`，next send some  data to client in the process handler of the ChannelInputShutdownEvent event。

### Actual behavior

client in the `FIN_WAIT2` state **can not read and process these data** which is sended by server in the `CLOSE_WAIT` state.

### Expected behavior

client in the `FIN_WAIT2` state can read and process these data which is sended by server in the `CLOSE_WAIT` state.

### Steps to reproduce

```
 ServerBootstrap server = new ServerBootstrap();
 server.childOption(ChannelOption.ALLOW_HALF_CLOSURE, true)

 Bootstrap client = new Bootstrap();
 client.option(ChannelOption.SO_LINGER, 1)
```
### Minimal yet complete reproducer code (or URL to code)

[see unit test method testHalfClosureReceiveDataOnFinalWait2StateWhenSoLingerSet](https://github.com/huibinliupush/netty/blob/afd2de8d81146301465ebc62fa1b5675931dd950/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java)

### Netty version
4.1.72.Final

### PR DESCRIPTION
**Pull request by huibinliupush**
Motivation:

we can not call `doDeregister` to cancel the key of the channel from the selector in the `io.netty.channel.AbstractChannel.AbstractUnsafe#shutdownOutput(io.netty.channel.ChannelPromise, java.lang.Throwable)` method。

- because we should ensure the side which enbale `SO_LINGER` and call `showdownOutput` to start TCP half-closure in fin_wait2 state can still receive and process the data which is send by another side in the close_wait state。

- and the shutdown function does not block regardless of the SO_LINGER setting on the socket,so we don't need to use GlobalEventExecutor to execute the shutdown

Modification:

In summary there is no need to call `prepareToClose()` in the `AbstractChannel.AbstractUnsafe#shutdownOutput` method

Result:

Fixes issue #11981 

client in the FIN_WAIT2 state can read and process these data which is sended by server in the CLOSE_WAIT state，when SO_LINGGER is used in the TCP half-closed scenario

### CODE REVIEW

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 1]
```diff
@@ -27,12 +27,20 @@
...
 import io.netty.channel.socket.ChannelInputShutdownEvent;
 import io.netty.channel.socket.ChannelInputShutdownReadComplete;
 import io.netty.channel.socket.ChannelOutputShutdownEvent;
 import io.netty.channel.socket.DuplexChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.serialization.ClassResolvers;
+import io.netty.handler.codec.serialization.ObjectDecoder;
+import io.netty.handler.codec.serialization.ObjectEncoder;
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:43]: Can we just use no codec for this test ?

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 2]
```diff
@@ -41,16 +49,98 @@
...
-import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assumptions.assumeFalse;
+import static java.util.concurrent.TimeUnit.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.*;
 
 public class SocketHalfClosedTest extends AbstractSocketTest {
+
+    @Test
+    public void testHalfClosureReceiveDataOnFinalWait2StateWhenSoLingerSet() throws Throwable {
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:63]: Please follow the same structure for tests as the rest of the file. This ensures the test will be run for NIO,OIO,Epoll and Kqueue. I suspect the same fix will need to go there as well.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 3]
```diff
@@ -41,16 +49,98 @@
...
+            ServerBootstrap sb = new ServerBootstrap();
+            sb.group(servergroup)
+              .channel(NioServerSocketChannel.class)
+              .childOption(ChannelOption.SO_LINGER, 1)
+              .childHandler(new ChannelInitializer<Channel>() {
+
+                  @Override
+                  protected void initChannel(Channel ch) throws Exception {
+                      ch.pipeline().addLast(new ObjectEncoder())
+                        .addLast(new ObjectDecoder(ClassResolvers.cacheDisabled(null)))
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:80]: Like said above there is no need to use a codec here.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 4]
```diff
@@ -41,16 +49,98 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.*;
 
 public class SocketHalfClosedTest extends AbstractSocketTest {
+
+    @Test
+    public void testHalfClosureReceiveDataOnFinalWait2StateWhenSoLingerSet() throws Throwable {
+        Channel serverChannel = null;
+        int PORT = 8005;
+        final AtomicBoolean clientReceiveDataOnFinalWait2 = new AtomicBoolean(false);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:66]: Use an CountdownLatch this way you can also remove the `Thread.sleep(...)` below.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 5]
```diff
@@ -41,16 +49,98 @@
...
+                                    ctx.close();
+                                }
+                            }
+                        });
+                  }
+              });
+
+            serverChannel = sb.bind(PORT).sync().channel();
+            Channel clientChannel = cb.connect("127.0.0.1",PORT).sync().channel();
+            Thread.sleep(1000);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:132]: See above

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 6]
```diff
@@ -41,16 +49,98 @@
...
+              .handler(new ChannelInitializer<Channel>() {
+                  @Override
+                  protected void initChannel(Channel ch) throws Exception {
+                      ch.pipeline().addLast(new ObjectEncoder())
+                        .addLast(new ObjectDecoder(ClassResolvers.cacheDisabled(null)))
+                        .addLast(new ChannelInboundHandlerAdapter(){
+
+                            @Override
+                            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                            }
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:114]: Call `ReferenceCountUtil.release(msg);`

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 7]
```diff
@@ -41,16 +49,98 @@
...
+                            public void channelActive(final ChannelHandlerContext ctx) {
+                                SocketChannel channel = (SocketChannel)ctx.channel();
+                                channel.shutdownOutput();
+
+                            }
+
+                            @Override
+                            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                                clientReceiveDataOnFinalWait2.set(true);
+                            }
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:93]: Call `ReferenceCountUtil.release(msg);`

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 8]
```diff
@@ -41,16 +49,98 @@
...
+
+            Bootstrap cb = new Bootstrap();
+            cb.group(clientgroup)
+              .channel(NioSocketChannel.class)
+              .option(ChannelOption.ALLOW_HALF_CLOSURE, true)
+              .handler(new ChannelInitializer<Channel>() {
+                  @Override
+                  protected void initChannel(Channel ch) throws Exception {
+                      ch.pipeline().addLast(new ObjectEncoder())
+                        .addLast(new ObjectDecoder(ClassResolvers.cacheDisabled(null)))
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:109]: Like said above remove the extra codecs

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 9]
```diff
@@ -650,38 +650,19 @@ private void shutdownOutput(final ChannelPromise promise, Throwable cause) {
...
-                    // Execute the shutdown.
-                    doShutdownOutput();
-                    promise.setSuccess();
-                } catch (Throwable err) {
-                    promise.setFailure(err);
-                } finally {
-                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
-                }
+
+            // when a side enbale SO_LINGER and call showdownOutput to start TCP half-closure,we can not call doDeregister here
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java:654]:
```suggestion
            // When a side enables SO_LINGER and calls showdownOutput(...) to start TCP half-closure, we can not call doDeregister here
```

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 10]
```diff
@@ -650,38 +650,19 @@ private void shutdownOutput(final ChannelPromise promise, Throwable cause) {
...
+            // because we should ensure this side in fin_wait2 state can still receive and process the data which is send by another side in the close_wait state。
+            // See https://github.com/netty/netty/issues/11981
+            try {
+                //The shutdown function does not block regardless of the SO_LINGER setting on the socket,so we don't need to use
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java:658]:
```suggestion
                // The shutdown function does not block regardless of the SO_LINGER setting on the socket,so we don't need to use
```

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 11]
```diff
@@ -650,38 +650,19 @@ private void shutdownOutput(final ChannelPromise promise, Throwable cause) {
...
+                //GlobalEventExecutor to execute the shutdown
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java:659]:
```suggestion
                // GlobalEventExecutor to execute the shutdown
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
A few comments :)

**Comment by normanmaurer** [comment]
@huibinliupush did you sigh our ICLA yet ?

**Comment by huibinliupush** [comment]
> @huibinliupush did you sigh our ICLA yet ?

yeah  i I already signed it before

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 12]
```diff
@@ -41,16 +44,104 @@
...
+
+                  @Override
+                  protected void initChannel(Channel ch) throws Exception {
+                      ch.pipeline().addLast(new  ChannelInboundHandlerAdapter(){
+
+                            @Override
+                            public void channelActive(final ChannelHandlerContext ctx) {
+                                SocketChannel channel = (SocketChannel)ctx.channel();
+                                channel.shutdownOutput();
+
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:94]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 13]
```diff
@@ -41,16 +44,104 @@
...
+        final AtomicBoolean clientReceiveDataOnFinalWait2 = new AtomicBoolean(false);
+        final CountDownLatch waitHalfClosureDone = new CountDownLatch(1);
+
+        final ByteBuf testHalfClosureSendMessage = Unpooled.buffer(16);
+        for (int i = 0; i < testHalfClosureSendMessage.capacity(); i ++) {
+            testHalfClosureSendMessage.writeByte((byte) i);
+        }
+
+        try {
+
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:82]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 14]
```diff
@@ -41,16 +44,104 @@
...
+                            @Override
+                            public void channelActive(final ChannelHandlerContext ctx) {
+                                SocketChannel channel = (SocketChannel)ctx.channel();
+                                channel.shutdownOutput();
+
+                            }
+
+                            @Override
+                            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                                clientReceiveDataOnFinalWait2.set(true);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:99]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 15]
```diff
@@ -41,16 +44,104 @@
...
+                                waitHalfClosureDone.countDown();
+                                ReferenceCountUtil.release(msg);
+                            }
+
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:103]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 16]
```diff
@@ -41,16 +44,104 @@
...
+        });
+    }
+
+
+    private void testHalfClosureReceiveDataOnFinalWait2StateWhenSoLingerSet(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+
+        Channel serverChannel = null;
+        Channel clientChannel = null;
+
+        final AtomicBoolean clientReceiveDataOnFinalWait2 = new AtomicBoolean(false);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:73]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 17]
```diff
@@ -41,16 +44,104 @@
 
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:47]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 18]
```diff
@@ -41,16 +44,104 @@
...
+                                }
+                            }
+                        });
+                  }
+              });
+
+            serverChannel = sb.bind().sync().channel();
+            clientChannel = cb.connect(serverChannel.localAddress()).sync().channel();
+            waitHalfClosureDone.await();
+            assertTrue(clientReceiveDataOnFinalWait2.get());
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:133]:
```suggestion
```

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java [hunk 19]
```diff
@@ -41,16 +44,104 @@
...
 import java.util.concurrent.atomic.AtomicReference;
 
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assumptions.assumeFalse;
+import static java.util.concurrent.TimeUnit.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.*;
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:53]: Please don't use wildcard imports

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 20]
```diff
@@ -650,38 +650,19 @@ private void shutdownOutput(final ChannelPromise promise, Throwable cause) {
...
-                } catch (Throwable err) {
-                    promise.setFailure(err);
-                } finally {
-                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
-                }
+
+            // When a side enables SO_LINGER and calls showdownOutput(...) to start TCP half-closure, we can not call doDeregister here
+            // because we should ensure this side in fin_wait2 state can still receive and process the data which is send by another side in the close_wait state。
+            // See https://github.com/netty/netty/issues/11981
+            try {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java:657]: @huibinliupush I am sure you will need to have the same change in the other transports as well (Epoll, KQueue).

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 20 above)
Comment (huibinliupush) [reply transport/src/main/java/io/netty/channel/AbstractChannel.java:657]: OK, i'll check it  now

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java (hunk 19 above)
Comment (huibinliupush) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:52]: really sorry for this

Comment (huibinliupush) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:52]: > Please don't use wildcard imports

really sorry for this

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 20 above)
Comment (huibinliupush) [reply transport/src/main/java/io/netty/channel/AbstractChannel.java:657]: > @huibinliupush I am sure you will need to have the same change in the other transports as well (Epoll, KQueue).

@normanmaurer OK, i'll check it right now

Comment (huibinliupush) [reply transport/src/main/java/io/netty/channel/AbstractChannel.java:657]: @normanmaurer 

i have checked `EpollSocketChannel` and `KQueueSocketChannel` again，they all call 
`((AbstractUnsafe) unsafe()).shutdownOutput(promise)` method to shutdownOutput In their parent class

Am I missing something?

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/AbstractChannel.java:657]: ah ok... Sorry I thought they also offload :)

**Comment by normanmaurer** [comment]
@huibinliupush please fix check style error:

```
Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:654: Line is longer than 120 characters (found 135). [LineLength]
901
Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:655: Line is longer than 120 characters (found 162). [LineLength]
902
Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:658: Line is longer than 120 characters (found 129). [LineLength]
```

**Comment by huibinliupush** [comment]
> @huibinliupush please fix check style error:
> 
> ```
> Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:654: Line is longer than 120 characters (found 135). [LineLength]
> 901
> Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:655: Line is longer than 120 characters (found 162). [LineLength]
> 902
> Error:  /home/runner/work/netty/netty/transport/src/main/java/io/netty/channel/AbstractChannel.java:658: Line is longer than 120 characters (found 129). [LineLength]
> ```

@normanmaurer  fix it

**Comment by normanmaurer** [comment]
@huibinliupush I think we are almost there..

Please also fix these:

```
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:65: two or more consecutive empty lines [RegexpMultiline]
5205
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:69: Line is longer than 120 characters (found 128). [LineLength]
5206
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:87:80: '{' is not preceded with whitespace. [WhitespaceAround]
5207
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:91:71: 'typecast' is not followed by whitespace. [WhitespaceAfter]
5208
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:101: an empty line before } [RegexpMultiline]
5209
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:102: two or more consecutive empty lines [RegexpMultiline]
5210
Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:110:79: '{' is not preceded with whitespace. [WhitespaceAround]
```

**Comment by huibinliupush** [comment]
> two or more consecutive empty lines [RegexpMultiline]

**Comment by huibinliupush** [comment]
> @huibinliupush I think we are almost there..
> 
> Please also fix these:
> 
> ```
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:65: two or more consecutive empty lines [RegexpMultiline]
> 5205
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:69: Line is longer than 120 characters (found 128). [LineLength]
> 5206
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:87:80: '{' is not preceded with whitespace. [WhitespaceAround]
> 5207
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:91:71: 'typecast' is not followed by whitespace. [WhitespaceAfter]
> 5208
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:101: an empty line before } [RegexpMultiline]
> 5209
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:102: two or more consecutive empty lines [RegexpMultiline]
> 5210
> Error:  /home/runner/work/netty/netty/testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketHalfClosedTest.java:110:79: '{' is not preceded with whitespace. [WhitespaceAround]
> ```

@normanmaurer   my intellij checksytle plugin reported  error but did not prompt specific error information, so it took me some time，sorry about that

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
@huibinliupush thanks a lot !

**Comment by huibinliupush** [comment]
> @huibinliupush thanks a lot !

@normanmaurer  so grateful to have learned so much from you，also thank you very much for reviewing this PR
